concordium-consensus/src/Concordium/Skov/Update.hs

```diff
@@ -479,87 +480,116 @@ doStoreBlock pb@GB.PendingBlock{..} = unlessShutDown $ do
-        _ -> snd <$> doReceiveTransactionInternal tr slot
-    when (ur == ResultSuccess) $ purgeTransactionTable False =<< currentTime
-    return ur
+  ur <- fst . snd <$> doReceiveTransactionInternal Single tr (utcTimeToTimestamp now) slot
+  when (ur == ResultSuccess) $ purgeTransactionTable False =<< currentTime  
+  return ur
 
-    where
-          isExpiryTooLate now = do
-            maxTimeToExpiry <- rpMaxTimeToExpiry <$> getRuntimeParameters
-            let expiry = msgExpiry tr
-            return $ expiry > maxTimeToExpiry + utcTimeToTransactionTime now
+-- |Type for describing the origin of the transaction.
+-- The transaction can either arrive at the consensus as a single transaction,
+-- or the transaction can be received as part of a block.
+data TransactionOrigin = Single | Block deriving (Eq, Show)
+
+-- |Type for describing if the transaction was accepted (and put into the transaction table)
+-- or if the transaction was rejected and not put into the table.
+-- This type exists because there is a difference in the 'tolerance' of accepting
+-- a transaction based on the transactions 'TransactionOrigin'.
+-- This is purposely not exposed in the 'SkovMonad' type class, in particular not in the 'receiveTransaction' function.
+-- as all other 'UpdateResult's than 'ResultSuccess' implies it has been rejected.
+data TransactionReceiveOutcome = Accepted | Rejected deriving (Eq, Show)
 
 -- |Add a transaction to the transaction table.  The 'Slot' should be
 -- the slot number of the block that the transaction was received with.
 -- This function should only be called when a transaction is received as part of a block.
 -- The difference from the above function is that this function returns an already existing
 -- transaction in case of a duplicate, ensuring more sharing of transaction data.
-doReceiveTransactionInternal :: (TreeStateMonad pv m) => BlockItem -> Slot -> m (Maybe BlockItem, UpdateResult)
-doReceiveTransactionInternal tr slot =
-        addCommitTransaction tr slot >>= \case
+-- This function also verifies the incoming transactions and adds them to the internal
+-- transaction verification cache such that the verification result can be used by the 'Scheduler'.
+-- The @origin@ parameter means if the transaction was received individually or as part of a block.
+-- The function returns the 'BlockItem' if it was "successfully verified" and added to the transaction table.
+-- Note. "Successfully verified" depends on the 'TransactionOrigin', see 'definitelyNotValid' below for details.
+doReceiveTransactionInternal :: (TreeStateMonad pv m) => TransactionOrigin -> BlockItem -> Timestamp -> Slot -> m (Maybe BlockItem, (UpdateResult, TransactionReceiveOutcome))
+doReceiveTransactionInternal origin tr ts slot = do
+   -- Don't accept the transaction if it was expired
+   if transactionExpired (msgExpiry tr) ts then return (Nothing, (ResultStale, Rejected))
+    else do
+      bs <- blockState =<< getFocusBlock
```

Comment:
I think we should use either the best or last finalized block instead of the focus block here although it mostly doesn't matter operationally. The focus block is an implementation detail, whereas the other two are well-defined from a protocol side.

Additionally, when we receive a transaction as part of a block, that block has a parent and we possibly already have the state of that block. If we do, we could use that here to be as precise as possible.